### ISSUE TITLE
Vincent's m68k-atari-mintelf-gcc -print-sysroot has wrong value

### ISSUE THREAD

**Issue by mikrosk** [freemint/m68k-atari-mint-gcc#38]
```
mikro@SK-WS-SROS1490:~/build-scripts$ m68k-atari-mintelf-gcc -print-sysroot
/build/gcc-m68k-atari-mintelf-FSofwz/gcc-m68k-atari-mintelf-13.2.0-mintelf-20240130/build/sys-root
```
It should be the path where gcc gets installed (/usr/m68k-atari-mintelf), not where it was built. :)

**Comment by benjihan** [comment]
I already told that to Vincent. I have a set of script that build the whole toolchain with a proper sysroot.
```
$ /usr/m68k-atari-mintelf/bin/m68k-atari-mintelf-gcc -print-sysroot
/usr/m68k-atari-mintelf/bin/../m68k-atari-mintelf/sys-root
```

**Comment by benjihan** [comment]
Also on a slightly related subject. What should be in your opinion the structure inside the sysroot. Currently I'm using the default `sys-root/usr`

**Comment by th-otto** [comment]
> Also on a slightly related subject. What should be in your opinion the structure inside the sysroot. Currently I'm using the default `sys-root/usr`

Yes, the sys-root directory should have the same structure as a native root filesystem, with subdirectories ``/usr``, ``/etc`` etc.

**Comment by vinriviere** [comment]
Indeed, my sysroot is messy. I had big trouble to understand how it was supposed to work. I will fix it.
However, unless I'm mistaken, the current ugly sysroot should not cause actual trouble.

Note:  I indeed used the sysroot for Ubuntu binaries, but not yet for Cygwin ones.

**Comment by benjihan** [comment]
It might be a problem in some cases. What if someone wants to install stuff in the sys-root ?

As far as I remember you just have to use `--with-sysroot` when configuring binutils and gcc.

- Configure, compile & install binutils
- Configure compile xgcc
- Configure, compile and install mintlib in sysroot
- Configure compile install fdlibm in sysroot
- Finish gcc compilation and install
 
I can share my build scripts.

**Comment by mikrosk** [comment]
For me it causes trouble when installing software: I have m68k-atari-mint-gcc and m68k-atari-mintelf-gcc installed side by side and to avoid changing --prefix paths on many places I change just `CC` and then use `$(CC) -print-sysroot` for determining the installation path.

**Comment by mikrosk** [comment]
Btw specifying `--with-sysroot=${INSTALL_DIR}/${TARGET}/sys-root` is usually enough. `--with-build-sysroot` is useful, too but I use it only for native builds.

**Comment by mikrosk** [comment]
Oh and my report is from WSL's ubuntu.

**Comment by vinriviere** [comment]
(remark below isn't an excuse for wrong sysroot in my Ubuntu builds)

> use `$(CC) -print-sysroot` for determining the installation path.

man gcc
```
       -print-sysroot
           Print  the  target  sysroot  directory that is used during compilation.  This is the target sysroot specified either at configure time or using the --sysroot option,
           possibly with an extra suffix that depends on compilation options.  If no target sysroot is specified, the option prints nothing.

```
So this doesn't seem to be a reliable technique to get the gcc installation directory.

Generally, I use something like this in my build scripts to dynamically determine the target directory:

```
$ realpath -m $(m68k-atari-mintelf-gcc -print-prog-name=as)/../..
/opt/cross-mintelf/m68k-atari-mintelf

```

**Comment by th-otto** [comment]
> So this doesn't seem to be a reliable technique to get the gcc installation directory.

IMHO this is the *only* reliable way to get the installation directory, since it will also work when you installed the toolchain to a different directory that was specified at configure time.

> $ realpath -m $(m68k-atari-mintelf-gcc -print-prog-name=as)/../..
> /opt/cross-mintelf/m68k-atari-mintelf

That only works if you configured with --prefix=/opt/cross-mintelf. For the old toolchain that will not work:

```
realpath -m $(/usr/bin/m68k-atari-mint-gcc -print-prog-name=as)/../..
/usr
```

But /usr/sys-root ist not the sys-root of that toolchain. That is```/usr/m68k-atari-mint/sys-root```

**Comment by benjihan** [comment]
But it does not really print the sys-root used  during compilation time. It changes when the toolchain is relocated as long as it  was inside the prefix tree. In other words it's relative to the bin when possible.

Anyway it's certainly cleaner to have the right sys-root printed.

**Comment by th-otto** [comment]
> But it does not really print the sys-root used during compilation time. It changes when the toolchain is relocated as long as it was inside the prefix tree.

Yes, that's the purpose. The configured sys-root would be rather useless if the toolchain is installed somewhere else.

**Comment by vinriviere** [comment]
(again: I will fix the wrong sysroot. But the discussion is interesting)

> That only works if you configured with --prefix=/opt/cross-mintelf. For the old toolchain that will not work:

Strangely it works as expected on my Cygwin setup with the old toolchain:

```
$ realpath -m $(m68k-atari-mint-gcc -print-prog-name=as)/../..
/opt/cross-mint/m68k-atari-mint
```

My intention was more something like this:
```
$ realpath -m $(type -p m68k-atari-mint-gcc)/../..
/opt/cross-mint
```
But in that case, the target directory `m68k-atari-mint` isn't included, so it has to be guessed from the toolchain prefix. Worse solution.

BTW, I'm not familiar with relocated toolchains (where actual installation directory isn't the compile-time prefix). The main purpose of the sysroot is to specify where the target include/lib are stored, instead of the classic $compilation_prefix/m68k-atari-mint? So this allows putting binaries (gcc, as...) in a different directory tree than include/libraries?

Sorry to insist, but I would like to deeply understand sysroot usage.
I never cared about sysroot in my old builds. I.e. `m68k-atari-mint-gcc -print-sysroot` returns nothing, and this never caused trouble.

**Comment by mikrosk** [comment]
> The main purpose of the sysroot is to specify where the target include/lib are stored, instead of the classic $compilation_prefix/m68k-atari-mint?

That's my understanding, yes. IMHO the sentence "Print  the  target  sysroot  directory" implies that you know where you want to install (as in from a package by package manager) it and at least _that_ should always work. So in your case `/opt/m68k-atari-mintelf` (but I'd suggest to comply with the new way, i.e. something like `/opt/m68k-atari-mintelf/sys-root/usr`). Otherwise the option `--with-build-sysroot` wouldn't be necessary.

**Comment by benjihan** [comment]
My understanding (I can be wrong) is that it's fine to have `-print-sysroot` returning nothing. it just indicates the compiler was build without sysroot support. It can be annoying if you rely on it but that's it. Otherwise when build with sysroot you can change it at compile time for whatever reason with `--sysroot dir`. As far as I can tell it's a convenience as you should be able to do the same with a more complicated command line.

Also gcc build documentation refers other methods as deprecated in favour of sys-root.

**Comment by th-otto** [comment]
>  So this allows putting binaries (gcc, as...) in a different directory tree than include/libraries?

No, the binaries have nothing to do with sysroot.

> Sorry to insist, but I would like to deeply understand sysroot usage. I never cared about sysroot in my old builds. I.e. `m68k-atari-mint-gcc -print-sysroot` returns nothing, and this never caused trouble.

Its main purpose is imho to simplify installation of libraries that are built later. But that depends on the scripts you are using for them. If someone wants to compile & install libraries locally that are not already available, he could use the `-print-sysroot` option to decide the installation directory. But when preparing archives that can be used by others, you have to use the same prefix as used for the toolchain anyway, because you cannot know where ***others*** have installed their toolchain.

However, there is some other trick that only works when building with -sysroot. You could install the binaries (gcc, as ld etc). in a completely different directory, but install the libraries/includes in the sys-root directory that was specified at configure time. I use that in my environment to install all the toolchains (gcc-4, gcc-7 etc) somewhere in my home directory in parallel, but have only one installation of mintlib etc. in /usr/m68k-atari-mint/sys-root. So if mintlib changes, i only have to update a single directory instead of dozens of them.
